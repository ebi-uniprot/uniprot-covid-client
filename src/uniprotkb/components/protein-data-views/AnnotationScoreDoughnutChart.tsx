import React, { FC } from 'react';
import { DoughnutChart } from 'franklin-sites';

export const annotationScoreToBin = (annotationScore: number) => {
  /*
    0-19  | bin 1
    20-39 | bin 2
    40-59 | bin 3
    60-79 | bin 4
    >=80  | bin 5
  */
  const bin = Math.floor(annotationScore / 20) + 1;
  return Math.min(bin, 5);
};

export enum DoughnutChartSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

type AnnotationScoreDoughnutChartProps = {
  score: number;
  size?: DoughnutChartSize;
};

const AnnotationScoreDoughnutChart: FC<AnnotationScoreDoughnutChartProps> = ({
  score,
  size = DoughnutChartSize.medium,
}) => {
  const annotationScoreBin = annotationScoreToBin(score);
  return (
    <span title="Annotation Score">
      <DoughnutChart percent={annotationScoreBin * 20} size={size}>
        {`${annotationScoreBin}/5`}
      </DoughnutChart>
    </span>
  );
};

export default AnnotationScoreDoughnutChart;
