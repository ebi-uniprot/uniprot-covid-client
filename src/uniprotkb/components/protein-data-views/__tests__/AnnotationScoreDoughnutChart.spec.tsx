import React from 'react';
import { render } from '@testing-library/react';
import AnnotationScoreDoughnutChart, {
  annotationScoreToBin,
} from '../AnnotationScoreDoughnutChart';

describe('AnnotationScoreDoughnutChart component', () => {
  test('should render', () => {
    const { asFragment } = render(<AnnotationScoreDoughnutChart score={50} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('annotationScoreToBin function', () => {
  /*
    0-19  | bin 1
    20-39 | bin 2
    40-59 | bin 3
    60-79 | bin 4
    >=80  | bin 5
  */
  const annotationScoreToBinTestPoints = [
    // [annotation score, bin]
    [0, 1],
    [1, 1],
    [19, 1],
    [20, 2],
    [39, 2],
    [40, 3],
    [59, 3],
    [60, 4],
    [79, 4],
    [80, 5],
    [81, 5],
    [1000, 5],
  ];
  annotationScoreToBinTestPoints.forEach(([annotationScore, bin]) => {
    test(`Should convert annotationScore of ${annotationScore} to bin ${bin}`, () => {
      expect(annotationScoreToBin(annotationScore)).toEqual(bin);
    });
  });
});
