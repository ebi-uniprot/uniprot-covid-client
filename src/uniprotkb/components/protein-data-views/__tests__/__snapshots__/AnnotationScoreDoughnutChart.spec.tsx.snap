// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationScoreDoughnutChart component should render 1`] = `
<DocumentFragment>
  <span
    title="Annotation Score"
  >
    <span
      class="doughnut-chart--medium colour-platinum"
    >
      <span
        class="doughnut-chart--medium__left-wrap"
      >
        <span
          class="doughnut-chart--medium__left-wrap__loader colour-sea-blue"
          style="transform: rotate(36deg);"
        />
      </span>
      <span
        class="doughnut-chart--medium__right-wrap"
      >
        <span
          class="doughnut-chart--medium__right-wrap__loader colour-sea-blue"
          style="transform: rotate(180deg);"
        />
      </span>
      <span
        class="doughnut-chart--medium__inner-circle"
      >
        3/5
      </span>
    </span>
  </span>
</DocumentFragment>
`;
